subprojects/core/src/test/groovy/org/gradle/internal/classpath/GroovyDynamicDispatchingInterceptingTest.groovy

```diff
@@ -138,6 +138,35 @@ class GroovyDynamicDispatchingInterceptingTest extends AbstractCallInterceptionT
         modifiedMetaClasses.every()
     }
 
+    def 'closure does not erroneously become effectively instrumented after it threw an exception'() {
+        given:
+        def transformedThrowingClosure = instrumentedClasses.instrumentedClosure {
+            throw new RuntimeException()
+        }
+        transformedThrowingClosure.delegate = new InterceptorTestReceiver()
+        def transformedInterceptedClosure = instrumentedClasses.instrumentedClosure {
+            test()
+        }
+        transformedInterceptedClosure.delegate = new FalseInterceptorTestReceiver()
+
+        when: 'a closure throwing an exception is called, then a closure that hits an instrumented call is called'
+        try {
+            transformedThrowingClosure()
+            // this calls throws an exception, but we expect the closure to be still correctly removed from control flow tracking
+        } catch (RuntimeException ignored) {}
+        transformedInterceptedClosure()
+
+        then: 'the closure that threw an exception should not have become effectively instrumented'
+        !(GroovySystem.metaClassRegistry.getMetaClass(InterceptorTestReceiver) instanceof CallInterceptingMetaClass)
```

Comment:
💭  Could we use Groovy 3's `!instanceof` operator to avoid extra parenthesis?

Reply:
Replacing this line with 
```groovy
GroovySystem.metaClassRegistry.getMetaClass(InterceptorTestReceiver) !instanceof CallInterceptingMetaClass
```
leads to an exception in Groovy compilation:
>```
> Unrecoverable compilation error: startup failed:
>  General error during instruction selection: RHS of the instanceof keyword must be a class name, not: org.codehaus.groovy.ast.expr.MethodCallExpression@1e715c75[object: org.codehaus.groovy.ast.expr.VariableExpression@370e8142[variable: $spock_valueRecorder type: org.spockframework.runtime.ValueRecorder] method: org.codehaus.groovy.ast.expr.ConstantExpression@315ec472[record] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@14703304[org.codehaus.groovy.ast.expr.MethodCallExpression@f6570b[object: org.codehaus.groovy.ast.expr.VariableExpression@4331c2c6[variable: $spock_valueRecorder type: org.spockframework.runtime.ValueRecorder] method: org.codehaus.groovy.ast.expr.ConstantExpression@48bf386[startRecordingValue] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@11f60800[org.codehaus.groovy.ast.expr.ConstantExpression@230a5b3a[6]]], org.codehaus.groovy.ast.expr.ClassExpression@6e608a6f[type: org.gradle.internal.classpath.CallInterceptingMetaClass]]]
>```
I'm not sure why, so I'll keep it the way it was.